packages/backend/src/core/BlockDownloader.ts

```diff
@@ -0,0 +1,218 @@
+} from '../peripherals/database/BlockRepository'
+import { EthereumClient } from '../peripherals/ethereum/EthereumClient'
+import { BlockNumber, BlockRange } from '../peripherals/ethereum/types'
+import { createEventEmitter } from '../tools/EventEmitter'
+import { JobQueue } from '../tools/JobQueue'
+import { Logger } from '../tools/Logger'
+
+export interface KnownBlock {
+  number: number
+  hash: string
+}
+
+type State = { t: 'not-started' } | { t: 'working'; lastKnownBlock: KnownBlock }
+
+export class BlockDownloader {
+  private events = createEventEmitter<BlockDownloaderEvents>()
+  private state: State = { t: 'not-started' }
+  private jobQueue: JobQueue
+
+  constructor(
+    private ethereumClient: EthereumClient,
+    private blockRepository: BlockRepository,
+    private logger: Logger
+  ) {
+    this.logger = this.logger.for(this)
+    this.jobQueue = new JobQueue({ maxConcurrentJobs: 1 }, this.logger)
+  }
+
+  async start() {
+    const lastKnownBlock = await this.blockRepository.getLast()
+
+    this.state = { t: 'working', lastKnownBlock }
+
+    return this.startBackgroundWork()
+  }
+
+  getStatus() {
+    const status: json = { status: this.state.t }
+
+    if (this.state.t === 'working')
+      status.lastKnownBlock = { ...this.state.lastKnownBlock }
+
+    return status
+  }
+
+  getLastKnownBlock() {
+    if (this.state.t !== 'working') {
+      throw new Error('Not started')
+    }
+    return this.state.lastKnownBlock
+  }
+
+  onNewBlocks(handler: (blockRange: BlockRange) => void) {
+    this.events.on('newBlocks', handler)
+    return () => {
+      this.events.off('newBlocks', handler)
+    }
+  }
+
+  onReorg(handler: (point: { firstChangedBlock: BlockNumber }) => void) {
+    this.events.on('reorg', handler)
+    return () => {
+      this.events.off('reorg', handler)
+    }
+  }
+
+  private startBackgroundWork() {
+    return this.ethereumClient.onBlock((block) =>
+      this.jobQueue.add({
+        name: `handleNewBlock-${block.number}-${block.hash}`,
+        execute: async () => this.handleNewBlock(block),
+      })
+    )
+  }
+
+  private async handleNewBlock(
+    latest: IncomingBlock,
+    retriesRemaining = 2
+  ): Promise<void> {
+    if (this.state.t !== 'working') return
+
+    await this.handlePastReorganizations()
+
+    const lastKnown = this.state.lastKnownBlock
+
+    // we fetch and save all blocks from the last known block,
+    // validating their parentHashes to ensure they form a valid chain
+    const newBlocks: IncomingBlock[] = await Promise.all(
+      range(lastKnown.number + 1, latest.number).map((blockNumber) =>
+        this.ethereumClient.getBlock(blockNumber)
+      )
+    )
+    newBlocks.push(latest)
+
+    // We fetch and save all blocks from the last known block, validating their
+    // parentHashes to ensure they form a valid chain, if they don't, we go back
+    // to the beginning of the procedure.
+    if (!isConsistentChain(newBlocks)) {
+      if (retriesRemaining > 0)
```

Comment:
Oh yeah, I wanted to talk about it.

So, currently jobQueue restarts all failed jobs.
- As inconsistent incoming blocks are something we expect, I don't want to throw an error on it — we expect it can happen, and we'll just try again.
- However, if it's not consistent after a few tries, I think we'd like to abort here, log an error that will land in an error reporting software and just forget about this job.

Reply:
I like the idea of a retry: false flag. However, what would be even more awesome is a maxRetries, then you can set it to 2 and not have any recursion.